### DISCUSSION TITLE
v6 : toDataURL keep controls

### DISCUSSION THREAD

**Discussion by Clemweb** [fabricjs/fabric.js#9837] [General]
Hi,

Is it normal that in v6 toDataURL return image with object controls ? I don't see it with v5.

I don't find any option to disabled this behavior (I just want the canvas with objects without controls on it)

![toDataUrl](https://github.com/fabricjs/fabric.js/assets/24622426/63bcffcd-3d95-4699-aa0c-5dcd902c31f2)

**Comment by ShaMan123** [comment]
Not normal, a bug

**Reply by ShaMan123** [reply]
Or you using after:render hooks?
Fabric abuses them so they are fired also when exporting to dataURL/canvas element

**Reply by Clemweb** [reply]
No I don't use this event, but I use another liek object:modified, object:scaling, ...

**Reply by ShaMan123** [reply]
I think we should have tests for this but I might be wrong

**Reply by asturur** [reply]
@ShaMan123 what do you mean that fabricJS abuses render hooks? As far as i can tell apart  emitting them doesn't do much with them

**Reply by ShaMan123** [reply]
It abuses the name to emit after:render for the main ctx, the top ctx, for canvas export and you don't have a decent way to distinguish between them

**Comment by asturur** [comment]
I see toDataURL calling toCanvasElement that calls renderCanvas, so for sure this is a bug that has been introduced and for which tests are not available or do not catch the regression.
